apps/next/app/dashboard/page.tsx

```diff
@@ -1,24 +1,100 @@
-import { redirect } from "next/navigation";
-import { auth, currentUser } from "@clerk/nextjs";
+import {
+  findCreateClerkOrganizationCreatedByUser,
+  mustGetClerkUser,
+} from "@/app/server/clerk";
 import { db } from "@/app/server/db";
-import { User, UserInsert, users } from "@/app/server/db/schema";
+import {
+  Team,
+  User,
+  UserInsert,
+  teams,
+  users,
+  usersToTeams,
+} from "@/app/server/db/schema";
 import { eq } from "drizzle-orm";
+import {
+  User as ClerkUser,
+  Organization as ClerkOrganization,
+} from "@clerk/nextjs/server";
+import { EnsureActiveOrgSetAndRedirect } from "@/components/EnsureActiveOrgSetAndRedirect";
+
+async function findCreateUserWithClerkId({
+  clerkId,
+  userInsert,
+}: {
+  clerkId: string;
+  userInsert: UserInsert;
+}): Promise<User> {
+  const user: User | null = await db
+    .select()
+    .from(users)
+    .where(eq(users.clerkId, clerkId))
+    .limit(1)
+    .then((rows) => rows[0] || null);
+  if (user) {
+    return user;
+  }
+  await db.insert(users).values(userInsert);
+  return findCreateUserWithClerkId({ clerkId, userInsert });
+}
+
+async function findCreateUserTeam({
+  teamName,
+  userId,
+  userClerkId,
+}: {
+  teamName: string;
+  userId: number;
+  userClerkId: string;
+}): Promise<Team> {
+  const userTeams = await db.query.teams.findMany({
+    where: (team, { eq }) => eq(team.name, teamName),
+    with: {
+      usersToTeams: {
+        where: (userToTeam, { eq }) => eq(userToTeam.userId, userId),
+      },
+    },
+  });
+  if (userTeams.length > 0) {
+    return userTeams[0];
+  }
+
+  // personal team not found, create it in Clerk
+  // then tx our db: insert team, insert usersToTeams
+  const clerkOrg: ClerkOrganization =
+    await findCreateClerkOrganizationCreatedByUser({
+      name: teamName,
+      createdByClerkId: userClerkId,
+    });
+  let team: Team | undefined;
+  await db.transaction(async (tx) => {
+    await tx.insert(teams).values({
+      clerkId: clerkOrg.id,
+      name: clerkOrg.name,
+      creatorId: userId,
+    });
+    team = await tx.query.teams.findFirst({
+      where: (team, { eq }) => eq(team.clerkId, clerkOrg.id),
+    });
+    if (!team) {
+      tx.rollback();
+      throw new Error("Team not found despite just creating it");
+    }
+    await tx.insert(usersToTeams).values({
```

Comment:
Does this mean a team can exist with its owner not actually belonging to it, until they're explicitly added?

Reply:
no, I don't think that's possible since the two operations (create team, associate user to team) are wrapped in a transaction

Reply:
Roger that.